### PR TITLE
Prioritize non-speculative tasks in scheduler and node allocator

### PR DESCRIPTION
**Pull request by losipiuk**
## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

**Comment by losipiuk** [comment]
CI: https://github.com/trinodb/trino/issues/17158

File: core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java [hunk 1]
```diff
@@ -645,12 +645,67 @@ public void testStressAcquireRelease()
...
+            NodeAllocator.NodeLease acquireSpeculative1 = nodeAllocator.acquire(REQ_NONE, DataSize.of(64, GIGABYTE), true);
+            assertAcquired(acquireSpeculative1, NODE_1);
+            NodeAllocator.NodeLease acquireSpeculative2 = nodeAllocator.acquire(REQ_NONE, DataSize.of(32, GIGABYTE), true);
+            assertAcquired(acquireSpeculative2, NODE_2);
+
+            // non-speculative tasks should still get node
+            NodeAllocator.NodeLease acquireNonSpeculative1 = nodeAllocator.acquire(REQ_NONE, DataSize.of(64, GIGABYTE), false);
+            assertAcquired(acquireNonSpeculative1, NODE_2);
+            NodeAllocator.NodeLease acquireNonSpeculative2 = nodeAllocator.acquire(REQ_NONE, DataSize.of(32, GIGABYTE), false);
+            assertAcquired(acquireNonSpeculative2, NODE_1);
```
Comment (linzebing) [inline core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java:671]: For my understanding, two nodes have a total of 128GB, but now we have 192GB scheduled (96GB speculative, 96GB non-speculative), right? How does this work out later, scheduler will preempt speculative tasks in favor of non-speculative tasks?

Comment (losipiuk) [reply core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java:671]: > How does this work out later, scheduler will preempt speculative tasks in favor of non-speculative tasks

They will work side by side until we run out of memory on node. In such case speculative task will be killed by low memory killer. If they will fit in available memory (task does not alway use all reserved memory) both should complete succesfully.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java [hunk 2]
```diff
@@ -2056,6 +2087,14 @@ public PrioritizedScheduledTask pollOrThrow()
             return prioritizedTask;
         }
 
+        public PrioritizedScheduledTask peekOrThrow()
+        {
+            IndexedPriorityQueue.Prioritized<ScheduledTask> task = queue.peekPrioritized();
+            checkState(task != null, "queue is empty");
+            // negate priority to reverse operation we do in addOrUpdate
+            return new PrioritizedScheduledTask(task.getValue(), toIntExact(-task.getPriority()));
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java:2097]: Yeah we can - will add commit.

**Review by arhimondr — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java:2097]: I will send separate PR

[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java:2097]: https://github.com/trinodb/trino/pull/17517

**Pull request merged**